File: python/diem2023.py
import requests
import json
import time
import threading
reqUrl = "https://hanoimoi.vn/api/getdiemthi2023"

headersList = {
 "authority": "hanoimoi.vn",
 "accept": "*/*",
 "accept-language": "en-US,en;q=0.9",
 "content-type": "application/x-www-form-urlencoded; charset=UTF-8",
 "cookie": "zone-430=0; zone-330=0; zone-230=0; zone-130=0",
 "origin": "https://hanoimoi.vn",
 "referer": "https://hanoimoi.vn/diem-thi-lop-10-2023",
 "sec-ch-ua-mobile": "?0",
 "sec-fetch-dest": "empty",
 "sec-fetch-mode": "cors",
 "sec-fetch-site": "same-origin",
 "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36 Edg/114.0.1823.58",
 "x-requested-with": "XMLHttpRequest" 
}

start = time.time()

def form1() :
    with open('data2.json', 'w', encoding='utf-8') as f:
        for i in range(200000, 205000):
            userid = str(i)
            payload = "t=2&q=" + userid
            response = requests.request("POST", reqUrl, data=payload,  headers=headersList)
            data = response.text
            json_object = json.loads(data)
            if json_object["SBD"] != "":
                json.dump(json_object, f, ensure_ascii=False, indent=4, separators=(", ", " : "))
                f.write(",\n")

def form2() :
    with open('data3.json', 'w', encoding='utf-8') as f:
        for i in range(205000, 210000):
            userid = str(i)
            payload = "t=2&q=" + userid
            response = requests.request("POST", reqUrl, data=payload,  headers=headersList)
            data = response.text
            json_object = json.loads(data)
            if json_object["SBD"] != "":
                json.dump(json_object, f, ensure_ascii=False, indent=4, separators=(", ", " : "))
                f.write(",\n")

def form3() :
    with open('data4.json', 'w', encoding='utf-8') as f:
        for i in range(210000, 215000):
            userid = str(i)
            payload = "t=2&q=" + userid
            response = requests.request("POST", reqUrl, data=payload,  headers=headersList)
            data = response.text
            json_object = json.loads(data)
            if json_object["SBD"] != "":
                json.dump(json_object, f, ensure_ascii=False, indent=4, separators=(", ", " : "))
                f.write(",\n")


# def form1():
#     with open('data2.json', 'w', encoding='utf-8') as f:
#         json_objects = [requests.get(reqUrl, params={"t": 2, "q": str(i)}).json() for i in range(100001, 105000)]
#         for json_object in json_objects:
#             if json_object["SBD"] != "":
#                 f.write(json_object["SBD"] + ",\n")

# def form2():
#     with open('data3.json', 'w', encoding='utf-8') as f:
#         json_objects = [requests.get(reqUrl, params={"t": 2, "q": str(i)}).json() for i in range(105000, 110000)]
#         for json_object in json_objects:
#             if json_object["SBD"] != "":
#                 f.write(json_object["SBD"] + ",\n")

# def form3():
#     with open('data4.json', 'w', encoding='utf-8') as f:
#         json_objects = [requests.get(reqUrl, params={"t": 2, "q": str(i)}).json() for i in range(110000, 115000)]
#         for json_object in json_objects:
#             if json_object["SBD"] != "":
#                 f.write(json_object["SBD"] + ",\n")

t1 = threading.Thread(target=form1)
t2 = threading.Thread(target=form2)
t3 = threading.Thread(target=form3)
t1.start()
t2.start()
t3.start()
t1.join()
t2.join()
t3.join()

# with concurrent.futures.ThreadPoolExecutor(3) as tp:
#     tp.submit(form1)
#     tp.submit(form2)
#     tp.submit(form3)

end = time.time()
print(f"Runtime of the program is {end - start}")
print("Done")

File: js/diem2023.js
const axios = require('axios').default;
const cheerio = require('cheerio');

const options = {
  method: 'GET',
  url: 'https://tsdaucap.hanoi.gov.vn/TraCuu/KetQuaTraCuuTuyenSinh10',
  params: {
    key: 'aW5wdXRERIaI5PYKLYXRhPTIweeg0mx4XsMTE2NSSzRjKXCGwZ0eXBlQ2hlY2s9MDI='
  },
  headers: {
    Accept: 'text/html, */*; q=0.01',
    'Accept-Language': 'en-US,en;q=0.9',
    Connection: 'keep-alive',
    Cookie: 'BIGipServerPool_TSDC_HN=1915228332.50719.0000; .AspNetCore.Antiforgery.68HoDSos0ic=CfDJ8CjI4AkC0xlNiBlQXi0796q7YGMIaMZsTi5RxX_GhTNJyU_H3LsN8GdE37fIf8Uar0YNRz_8IZ-IJioWmOD-aqvtqZVlyLnQoHdxas16GFlHVVE6SD7EpFI09Wsb0wLRYQ7HltmPPeK4DaB-QY5UDuw',
    Referer: 'https://tsdaucap.hanoi.gov.vn/tra-cuu-tuyen-sinh-10',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'same-origin',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36 Edg/114.0.1823.58',
    'X-Requested-With': 'XMLHttpRequest',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '^^"Windows^^"'
  }
};

axios.request(options)
  .then(function (response) {
    const $ = cheerio.load(response.data);
    const boxThongTinDiem = $('.col-md-8.box-thong-tin-diem');
    const rows = boxThongTinDiem.find('.row');

    // Extract the relevant data from the rows
    const soBaoDanh = rows.eq(0).find('b').text();
    const maHocSinh = rows.eq(1).find('b').text();
    const hoTen = rows.eq(2).find('b').text();
    const diemThi = rows.eq(3).find('b').text();

    // Do something with the extracted data
    console.log({
      soBaoDanh,
      maHocSinh,
      hoTen,
      diemThi,
    });
  })
  .catch(function (error) {
    console.error(error);
  });